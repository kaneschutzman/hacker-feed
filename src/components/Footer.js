import React from 'react';

const Footer = () => (
    <p className="center-align">made with <span role="img" style={{color:'red'}} aria-label="jsx-ally/accessible-emoji">❤️</span> by <a href="https://github.com/skywalker212" style={{'textDecoration':'none'}} rel="noopener noreferrer" target="_blank">akash</a>
    </p>
);

export default Footer;