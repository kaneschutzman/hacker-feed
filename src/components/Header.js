import React from 'react';

const Header = (props)=>{
    return (
        <div className="center-align">
            <h1 className="App-title">HackerFeed</h1>
            <h5 className="App-desc">Your own HackerNews feed</h5>
            <h6>click on the title to open link</h6>
        </div>
    );
};

export default Header;