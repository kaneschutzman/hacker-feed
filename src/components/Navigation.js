import React from 'react';
import {connect} from 'react-redux';
import {setStories, changeLoading} from '../actions/pages';
import changeStories from '../actions/changeResults';

export class Navigation extends React.Component{
    setTop = () => {
        this.props.changeLoading(true);
        changeStories(this.props.setStories, this.props.changeLoading,'top');
    }
    setBest = () => {
        this.props.changeLoading(true);
        changeStories(this.props.setStories,this.props.changeLoading,'best');
    }
    setNew = () => {
        this.props.changeLoading(true);
        changeStories(this.props.setStories,this.props.changeLoading,'new');
    }
    render= () =>  (<div className="center-align navigation">
                <button className="waves-effect waves-light btn-small z-depth-3" onClick={this.setTop}>Top Stories</button>
                <button className="waves-effect waves-light btn-small z-depth-2" onClick={this.setBest}>Best Stories</button>
                <button className="waves-effect waves-light btn-small z-depth-3" onClick={this.setNew}>New Stories</button>
            </div>);
}

const mapStateToProps = (state) => ({
    curr: state.curr
});

const mapDispatchToProps = (dispatch)=>({
    setStories: (stories,type) => dispatch(setStories(stories,type)),
    changeLoading: (type) => dispatch(changeLoading(type))
});

export default connect(mapStateToProps,mapDispatchToProps)(Navigation);