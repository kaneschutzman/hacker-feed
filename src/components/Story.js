import React from 'react';
import moment from 'moment';

export default class Story extends React.Component{
    constructor(){
        super();
        this.state={
            displayText: false
        }
    }
    componentDidMount(){
        this._isMounted=true;
        fetch('https://hacker-news.firebaseio.com/v0/item/'+this.props.id+'.json')
        .then(res=>res.json())
        .then(data=>{
            if(this._isMounted){
                this.setState((state)=>{
                    return {
                        state,
                        ...data
                    }
                });
            }
        }).catch(err=>{
            if(this._isMounted){
                this.setState({
                    title: err
                });
            }
        });
    }

    componentWillUnmount(){
        this._isMounted=false;
    }

    changeDisplayState = () =>{
        this.setState((state)=>{
            return {
                displayText: !state.displayText
            };
        });
    }

    render(){
        return (
            <div className="story card blue-grey darken-1 z-depth-1">
                <div className="card-content white-text">
                    {this.state.title && 
                    <div>
                        <a href={this.state.url} className="card-title light-green-text text-accent-3" target="_blank" style={{display:'inline'}}>{this.state.title}</a>
                        <p style={{display:'inline',marginLeft:'10px'}}>by {this.state.by}</p>
                    </div>}
                    <div>
                        {this.state.time && <div>Posted at: {moment.unix(this.state.time).format('MMM Do, YY')}</div>}
                        {this.state.score && <p>Score: {this.state.score}</p>}
                    </div>
                    {this.state.text && 
                    <div>
                        <button className="waves-effect waves-light btn-small" onClick={this.changeDisplayState}>{this.state.displayText?'Hide':'Show'} Text</button>
                        {this.state.displayText && <p>{this.state.text}</p>}
                    </div>
                    }
                </div>
            </div>
        );
    }
}