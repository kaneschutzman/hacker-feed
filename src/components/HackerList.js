import React from 'react';
import {connect} from 'react-redux';
import Story from './Story';
import {nextPage, prevPage, jumpToPage,setError} from '../actions/pages';

class HackerList extends React.Component{
    filterStories(){
        return this.props.stories
                .filter((data,index)=>index>=(this.props.page-1)*10 && index<(this.props.page-1)*10+10)
                .map(data=><Story key={data} id={data} />);
    }
    jumpTo = (event) => {
        const pageNo = event.target.value;
        if(!pageNo) this.props.setError(true);
        else if(pageNo>=1 && pageNo<=this.props.pages){
            this.props.jumpToPage(pageNo);
            this.props.setError(false);
        }
    }
    render(){
        return (
            <div>
                {this.props.error && <p>Please set a value for page</p>}
                {this.props.stories.length===0 && <p>Please wait...</p>}
                <h6 className='center-align' style={{margin:'20px 0'}}>{this.props.curr} stories</h6>
                {this.props.stories && this.filterStories()}
                {this.props.pages>0 && <div className="center-align">
                    <p>Showing page {this.props.page} of {this.props.pages}</p>
                    <button className="prev-btn black-text waves-effect waves-light btn-small yellow accent-2" disabled={this.props.page===1} onClick={this.props.prevPage}>Previous Page</button>
                    {/*<input type='text' onChange={this.jumpTo} value={this.props.page} />*/}
                    <button className="next-btn black-text waves-effect waves-light btn-small red accent-2" disabled={this.props.page===this.props.pages} onClick={this.props.nextPage}>NextPage</button>
                </div>}
            </div>
        );
    }
}

const mapStateToProps = (state) => ({page: state.page,pages: state.pages,stories: state.stories,error: state.error,curr:state.curr});
const mapDispatchToProps = (dispatch) => ({
    nextPage: () => dispatch(nextPage()),
    prevPage: () => dispatch(prevPage()),
    jumpToPage: (page) => dispatch(jumpToPage(page)),
    setError: (err) => dispatch(setError(err))
});

export default connect(mapStateToProps,mapDispatchToProps)(HackerList);