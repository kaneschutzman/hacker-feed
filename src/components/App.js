import React, { Component } from 'react';
import {connect} from 'react-redux';
import HackerList from './HackerList';
import Header from './Header';
import {setPages, setStories} from '../actions/pages';
import '../styles/main.css';

class App extends Component {
  componentDidMount(){
    fetch('https://hacker-news.firebaseio.com/v0/topstories.json')
        .then(result=>result.json())
        .then(data=>{
            this.props.setStories(data);
            this.props.setPages(Math.ceil(data.length/10));
        });
  }
  render() {
    return (
      <div className="App">
        <Header />
        <HackerList />
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  //function for future use
});

const mapDispatchToProps = (dispatch) => ({
  setPages: (pages) => dispatch(setPages(pages)),
  setStories: (stories) => dispatch(setStories(stories))
});

export default connect(mapStateToProps,mapDispatchToProps)(App);
