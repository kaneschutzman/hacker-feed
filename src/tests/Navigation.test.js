import React from 'react';
import {shallow} from 'enzyme';
import {Navigation} from '../components/Navigation';

it('renders Navigation component successfully',()=>{
    const setStory = jest.fn();
    const wrapper = shallow(<Navigation setStories={setStory}/>);
    expect(wrapper).toMatchSnapshot();
});