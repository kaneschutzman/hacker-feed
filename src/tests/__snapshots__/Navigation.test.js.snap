// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Navigation component successfully 1`] = `
<div
  className="center-align navigation"
>
  <button
    className="waves-effect waves-light btn-small z-depth-3"
    onClick={[Function]}
  >
    Top Stories
  </button>
  <button
    className="waves-effect waves-light btn-small z-depth-2"
    onClick={[Function]}
  >
    Best Stories
  </button>
  <button
    className="waves-effect waves-light btn-small z-depth-3"
    onClick={[Function]}
  >
    New Stories
  </button>
</div>
`;
